### ISSUE TITLE
High memory consumption during inference

### ISSUE THREAD

**Issue by MTDzi** [ultralytics/yolov5#4647]
## ❔Question

I noticed that *all* of the Python tutorials supplied in the [Nvidia Jetson EXPORT competition](https://github.com/ultralytics/yolov5/discussions/3426) consume a lot of memory (3GB or more), even with the smallest (yolov5s) model.

What the heck? Does anyone have an idea why that is?

The TensorRTx Cpp version, on the other hand,  consumes ~800MB at most. I also read [here](https://forums.developer.nvidia.com/t/high-ram-consumption-with-cuda-and-tensorrt-on-jetson-xavier-nx/183109/11) that the cause might be that the .engine file deserialization might require loading libraries that are not needed for inference (but they occupy the memory afterwards) but not sure, I just started looking into it. 

Any and all thoughts welcome!

**Comment by glenn-jocher** [comment]
@MTDzi when you say high memory consumption, high relative to what comparative baseline?

**Comment by MTDzi** [comment]
@glenn-jocher thanks for the reply!

TL;DR Python inference (I tested it using TensorRT, ONNX, and TorchScript) consumes much more memory as compared to Cpp inference (I tested only the TensorRTx version).

To clarify, and as a concrete example: when I run [this](https://github.com/wang-xinyu/tensorrtx/blob/master/yolov5/yolov5_trt.py) Python script my memory usage gets to ~3GB, but the equivalent Cpp version ([here](https://github.com/wang-xinyu/tensorrtx/blob/master/yolov5/yolov5.cpp)) is 800MB. The same model, the same underlying engine (TensorRT), seemingly the same procedure, yet the memory usage is much higher when using the Python version.

Basically, that Cpp version is the lowest-memory-consuming approach  that I found, and all Python versions (TensorRT, ONNX, TorchScript) consume almost 3x as much memory.

**Comment by glenn-jocher** [comment]
@MTDzi interesting, thanks for the results! 

Most CPU inference backends I think will be running full FP32 mode, so maybe quantization plays a part. Maybe also one of the main memory increasing aspects of modern AIs are the shortcut links that pass information from earlier in the model to later in the model. This means each shortcut output must also be stored into memory, but we do this in a smart way so only the required outputs are saved and not all outputs.

**Comment by MTDzi** [comment]
@glenn-jocher did you mean "Cpp" instead of "CPU"? Because both memory usages (Python and Cpp) were conducted using a GPU.

When doing inference on CPU the memory usage for the Python versions (using PyTorch, ONNX, and TorchScript) is low, I don't remember the exact numbers but definitely lower than 2GB.

If this helps in any way, I can record my screen and voice and upload it to YouTube (or wherever) so that I can better provide evidence for what I'm experiencing. What do you think?

**Comment by glenn-jocher** [comment]
@MTDzi unless you are seeing an error then everything is working correctly and there is no action for us to take.

**Comment by MTDzi** [comment]
@glenn-jocher I wanted to ask in the hopes that someone encountered this problem and maybe has a workaround (better than using the Cpp version).

Thanks for your help, I'll dig further at the TensorRTx project.

**Comment by Adrien-tt** [comment]
I’m interesting by this subject, I use yolov5s model and Python inference (TorchScript). I also note that my memory usage gets to 3-4GB.

If someone has an idea to reduce memory usage with any python version, I will take it !

**Comment by glenn-jocher** [comment]
@Adrien-tt have you tried TFLite inference? Maybe the 16 and 8 bit models consume less RAM. I haven't run this test myself.

**Comment by Adrien-tt** [comment]
I have tried TFLite inference with a half-precision model (FP16) and memory usage gets to 3-4GB too.

I see that I have lots of threads with my python script and I understand that it is for initialization but can we reduce this number or memory allocated for each thread?

**Comment by MTDzi** [comment]
@Adrien-tt I tried to follow this rabbit hole for a bit, and created an Issue in the tensorrtx repo [here](https://github.com/wang-xinyu/tensorrtx/issues/689).

But, honestly, the best I came up with is: using the Cpp version. I think the Python version doesn't give you full control over the memory needed by the TensorRT runtime and that's the root cause. But I don't know for sure.

**Comment by glenn-jocher** [comment]
@Adrien-tt interesting! Thank you for your feature suggestion on how to improve YOLOv5 🚀!

The fastest and easiest way to incorporate your ideas into the official codebase is to submit a Pull Request (PR) implementing your idea, and if applicable providing before and after profiling/inference/training results to help us understand the improvement your feature provides. This allows us to directly see the changes in the code and to understand how they affect workflows and performance.

Please see our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/) to get started.

**Comment by isomatical** [comment]
Hi, I met the same problem and had no idea about it. Any update on this issue?

**Comment by isomatical** [comment]
> 

UPDATE: I fixed the problem by using ctypes to wrap up the cpp api.

**Comment by MTDzi** [comment]
@isomatical whoa, awesome.
Do you have the option of sharing your solution?

**Comment by isomatical** [comment]
@MTDzi Sure, I will share it soon. (codes are a bit messy right now)

**Comment by MTDzi** [comment]
@isomatical Great, thanks!

**Comment by mohammad69h94** [comment]
Hello, I met the same problem. Does anyone have a suggestion?

**Comment by isomatical** [comment]
@MTDzi 
Sorry for the late reply! Codes are [here](https://github.com/isomatical/tensorrtx).

**Comment by MTDzi** [comment]
@isomatical no problemo, thanks, this is awesome!

**Comment by ankitvirla** [comment]
I'm also facing the same problem. Whenever I'm loading the yolov5 model on GPU it's also taking up to 4 GB of CPU RAM.
Did anyone solve that?

**Comment by H-deep** [comment]
Can anybody tell the cause of this problem?

**Comment by glenn-jocher** [comment]
@birla8319 @H-deep I think this may simply be due to loading torch into the environment. You might try running inference with a different deployment format to eliminate the torch dependency.

YOLOv5 🚀 offers export to most popular formats used today. See our [TFLite, ONNX, CoreML, TensorRT Export Tutorial](https://docs.ultralytics.com/yolov5/tutorials/model_export) for details.

## Formats

YOLOv5 inference is officially supported in 11 formats:

💡 ProTip: **TensorRT** may be up to 2-5X faster than PyTorch on [**GPU benchmarks**](https://github.com/ultralytics/yolov5/pull/6963)
💡 ProTip: **ONNX** and **OpenVINO** may be up to 2-3X faster than PyTorch on [**CPU benchmarks**](https://github.com/ultralytics/yolov5/pull/6613)

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`

## CPU Benchmarks on Colab Pro+ CPU instance

Full [CPU benchmarks](https://github.com/ultralytics/yolov5/pull/6613)

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=cpu, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CPU
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 41.5/166.8 GB disk)

Benchmarks complete (241.20s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623               127.61
1             TorchScript        0.4623               131.23
2                    ONNX        0.4623                69.34
3                OpenVINO        0.4623                66.52
4                TensorRT           NaN                  NaN
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623               123.79
7     TensorFlow GraphDef        0.4623               121.57
8         TensorFlow Lite        0.4623               316.61
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

## GPU Benchmarks on Colab Pro+ V100 instance

Full [GPU benchmarks](https://github.com/ultralytics/yolov5/pull/6963)

```
benchmarks: weights=/content/yolov5/yolov5s.pt, imgsz=640, batch_size=1, data=/content/yolov5/data/coco128.yaml, device=0, half=False, test=False
Checking setup...
YOLOv5 🚀 v6.1-135-g7926afc torch 1.10.0+cu111 CUDA:0 (Tesla V100-SXM2-16GB, 16160MiB)
Setup complete ✅ (8 CPUs, 51.0 GB RAM, 46.7/166.8 GB disk)

Benchmarks complete (458.07s)
                   Format  mAP@0.5:0.95  Inference time (ms)
0                 PyTorch        0.4623                10.19
1             TorchScript        0.4623                 6.85
2                    ONNX        0.4623                14.63
3                OpenVINO           NaN                  NaN
4                TensorRT        0.4617                 1.89
5                  CoreML           NaN                  NaN
6   TensorFlow SavedModel        0.4623                21.28
7     TensorFlow GraphDef        0.4623                21.22
8         TensorFlow Lite           NaN                  NaN
9     TensorFlow Edge TPU           NaN                  NaN
10          TensorFlow.js           NaN                  NaN
```

Good luck 🍀 and let us know if you have any other questions!
